crates/core/tedge_api/src/mqtt_topics.rs

```diff
@@ -305,6 +310,75 @@ impl EntityTopicId {
         }
         .map(|parent_id| EntityTopicId(format!("device/{parent_id}//")))
     }
+
+    /// If `self` is a device topic id, return a service topic id under this
+    /// device.
+    ///
+    /// The device topic id must be in a format: "device/DEVICE_NAME//"; if not,
+    /// `None` will be returned.
+    pub fn to_service_topic_id(&self, service_name: &str) -> Option<ServiceTopicId> {
+        if let ["device", device_name, "", ""] = self.0.split('/').collect::<Vec<&str>>()[..] {
+            return Some(ServiceTopicId(EntityTopicId(format!(
+                "device/{device_name}/service/{service_name}"
+            ))));
+        }
+        None
+    }
+
+    /// Returns an array of all segments of this entity topic.
+    pub fn segments(&self) -> [&str; ENTITY_ID_SEGMENTS] {
+        let mut segments = self.0.split('/');
+        let seg1 = segments.next().unwrap();
+        let seg2 = segments.next().unwrap();
+        let seg3 = segments.next().unwrap();
+        let seg4 = segments.next().unwrap();
+        [seg1, seg2, seg3, seg4]
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Represents an entity topic identifier known to be a service.
+///
+/// It's most often in a format `device/DEVICE_NAME/service/SERVICE_NAME`.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ServiceTopicId(EntityTopicId);
```

Comment:
I'm not convinced by this type.

Why would we have a specific type to identify services but not devices?

Reply:
Also addressed it [here](https://github.com/thin-edge/thin-edge.io/pull/2271#discussion_r1331351769).

We can have a specific type to identify devices as well, just by the scope of this PR I thought it was a good place to add one for a service. I think as we plan to support custom MQTT schemes, where any identifier can be either a device or a service, it makes sense to save this information by encoding it in a type. Granted, producers of these wrappers are currently not well defined, so maybe this is a premature abstraction, but ultimately, I think these `ServiceTopicId`/`DeviceTopicId` wrapper types could be produced by the entity store (that is, they can be derived from `EntityType` which is a part of `EntityMetadata`), or derived from a default topic scheme.

Reply:
Responded in [this comment](https://github.com/thin-edge/thin-edge.io/pull/2276/files/5df95385746516786409c99c7f279cf028e082d2#r1338516075), if that's your preference I can remove `ServiceTopicId` and use `EntityTopicId`.

Reply:
As of now, I see indeed little value in `ServiceTopicId` and `DeviceTopicId`. However, this is not blocking.

What really add values is the `struct Service` which combines the `EntityTopicId` of a service with the `EntityTopicId` of its parent device removing the need for guessing one from the other.

Reply:
I finally understood what is worrying me. This `struct ServiceTopicId` and even the `struct Service` are halfway between `EntityTopicId` and `EntityMetadata`.

- `EntityTopicId`s are just identifiers of sources & targets of MQTT messages - with no assumption of the nature of these sources and targets.
- `EntityMetadata`s describe these entities - notably providing their type and their parent if any.

So, if a feature needs to know the type of entity or its parent, then the implementation must be given an `EntityMetadata` not a decorated `EntityTopicId`.

Hence, now I do think that it would be better to not introduce neither `ServiceTopicId` nor `DeviceTopicId` and not even `Service`.

Reply:
> So, if a feature needs to know the type of entity or its parent, then the implementation must be given an `EntityMetadata` not a decorated `EntityTopicId`.

Yeah, I agree that `EntityMetadata` is probably the closest to what we eventually want to have. One problem with it though is that this type is designed for flat storage in a vector, i.e. `parent` field being just a string key that one can look in a hashmap - this is far simpler than using references. So currently, if we want to see who is the parent in `EntityMetadata`, we have to call `.get()` again on the entity store.

The solution for this I think would be to have an `Entity` struct, which is models the entity so as to make invalid states unrepresentable and contain any other relevant entities, e.g. a parent, so one wouldn't have to call entity store again. This is similar to `.ancestors()` method, but instead of returning IDs, it could return objects with all necessary information.

For now though, as agreed, I won't be making any sweeping API changes and just do it with `EntityMetadata` directly, in order to have full picture of the pain points before resolving them.